lwr_controllers/src/arm_state_controller.cpp

```diff
@@ -0,0 +1,121 @@
+        
+        // get publishing period
+        if (!nh_.getParam("publish_rate", publish_rate_)) {
+            ROS_ERROR("Parameter 'publish_rate' not set");
+            return false;
+        }
+        
+        realtime_pub_.reset(new realtime_tools::RealtimePublisher<lwr_controllers::ArmState>(nh_,"arm_state",4));
+        realtime_pub_->msg_.est_ext_torques.resize(kdl_chain_.getNrOfJoints());
+        
+        gravity_.reset(new KDL::Vector(0.0, 0.0, -9.81)); // TODO: compute from actual robot position (TF?)
+        id_solver_.reset(new KDL::ChainIdSolver_RNE(kdl_chain_, *gravity_));
+        jac_solver_.reset(new KDL::ChainJntToJacSolver(kdl_chain_));
+        fk_solver_.reset(new KDL::ChainFkSolverPos_recursive(kdl_chain_));
+        jacobian_.reset(new KDL::Jacobian(kdl_chain_.getNrOfJoints()));
+        joint_position_.reset(new KDL::JntArray(kdl_chain_.getNrOfJoints()));
+        joint_velocity_.reset(new KDL::JntArray(kdl_chain_.getNrOfJoints()));
+        joint_acceleration_.reset(new KDL::JntArray(kdl_chain_.getNrOfJoints()));
+        joint_wrenches_.reset(new KDL::Wrenches(kdl_chain_.getNrOfJoints()));
+        joint_effort_est_.reset(new KDL::JntArray(kdl_chain_.getNrOfJoints()));
+        
+        return true;		
+    }
+    
+    void ArmStateController::starting(const ros::Time& time)
+    {
+        last_publish_time_ = time;
+        for (unsigned i = 0; i < joint_handles_.size(); i++){
+            (*joint_position_)(i) = joint_handles_[i].getPosition();
+            (*joint_velocity_)(i) = joint_handles_[i].getVelocity();
+            (*joint_acceleration_)(i) = 0; 
+        }
+        ROS_INFO("started");
+    }
+    
+    void ArmStateController::update(const ros::Time& time, const ros::Duration& period)
+    {
+        // limit rate of publishing
+        if (publish_rate_ > 0.0 && last_publish_time_ + ros::Duration(1.0/publish_rate_) < time){
+            
+            if (realtime_pub_->trylock()) {
+                // we're actually publishing, so increment time
+                last_publish_time_ = last_publish_time_ + ros::Duration(1.0/publish_rate_);
+                
+                for (unsigned i = 0; i < joint_handles_.size(); i++) {
+                    float acceleration = filters::exponentialSmoothing((joint_handles_[i].getPosition() - (*joint_position_)(i))/period.toSec(), joint_handles_[i].getVelocity(), 0.2);
+                    (*joint_position_)(i) = joint_handles_[i].getPosition();
+                    (*joint_velocity_)(i) = joint_handles_[i].getVelocity();
+                    (*joint_acceleration_)(i) = acceleration; 
+                }
+                                
+                // Compute Dynamics 
+                int ret = id_solver_->CartToJnt(*joint_position_, 
+                                                *joint_velocity_, 
+                                                *joint_acceleration_, 
+                                                *joint_wrenches_,
```

Comment:
Ok. I'm still playing with it because today I get very crappy performance out of the controller. I don't know if I changed something wrong; in the meantime I realised that I was computing the acceleration in a very wrong way :innocent: so I fixed it